Core/BlockerListsLoader.swift

```diff
@@ -54,32 +54,38 @@ public class BlockerListsLoader {
         let urls = AppUrls()
 
         APIRequest(url: urls.disconnectMeBlockList).execute { (data, error) in
-            if let data = data {
+            Logger.log(items: "DisconnectMe request completion", "\(String(describing: error))")
+           return BlockerListsLoader.handleResponseAndSignalCompletion(data: data, error: error, semaphore: semaphore, dataHandler: { (data) in
                 try? DisconnectMeStore.shared.persist(data: data)
-            }
-            Logger.log(items: "DisconnectMeRequest", DisconnectMeStore.shared.allTrackers.count, "\(String(describing: error))")
-            semaphore.signal()
+            })
         }
 
         APIRequest(url: urls.easylistBlockList).execute { (data, error) in
-            if let data = data {
+            Logger.log(items: "Easylist request completion", "\(String(describing: error))")
+            return BlockerListsLoader.handleResponseAndSignalCompletion(data: data, error: error, semaphore: semaphore, dataHandler: { (data) in
                 self.easylistStore.persistEasylist(data: data)
-            }
-
-            Logger.log(items: "EasylistRequest", "\(String(describing: error))")
-            semaphore.signal()
+            })
         }
 
         APIRequest(url: urls.easylistPrivacyBlockList).execute { (data, error) in
-            if let data = data {
+            Logger.log(items: "EasylistPrivate request completion", "\(String(describing: error))")
+            return BlockerListsLoader.handleResponseAndSignalCompletion(data: data, error: error, semaphore: semaphore, dataHandler: { (data) in
                 self.easylistStore.persistEasylistPrivacy(data: data)
-            }
+            })
 
-            Logger.log(items: "EasylistPrivacyRequest", "\(String(describing: error))")
-            semaphore.signal()
         }
 
         return 3
     }
 
+    private class func handleResponseAndSignalCompletion(data: Data?, error: Error?, semaphore: DispatchSemaphore, dataHandler: (Data) -> Void) -> APIRequestCompleteionResult {
+        var result: APIRequestCompleteionResult = .errorHandled
```

Comment:
Why is the valid case of a cached response being treated as an error?

Reply:
I didn’t want to store he etag if the handler hadn’t done something with the data (reinforcing the data completion handler’s role regarding data storage), and I didn’t want to use a bool as that was ambiguous. Let me think about this one.